server/src/main/java/org/elasticsearch/index/query/TermsSetQueryBuilder.java

```diff
@@ -342,6 +353,109 @@ public LongValuesSource rewrite(IndexSearcher searcher) throws IOException {
+                // as far as this query is concerned
+                return o;
+            }
+        };
+    }
+
+    /**
+     * Convert the list in a way that optimizes storage in the case that all
+     * elements are either integers or {@link String}s/{@link BytesRef}/
+     * {@link CharBuffer}s. This is useful to help garbage collections for
+     * use-cases that involve sending very large terms queries to Elasticsearch.
+     * If the list does not only contain integers or {@link String}s, then a
+     * list is returned where all {@link String}/{@link CharBuffer}s have been
+     * replaced with {@link BytesRef}s.
+     */
+    static List<?> convert(List<?> list) {
+        if (list.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        final boolean allNumbers = list.stream().allMatch(o -> o != null && INTEGER_TYPES.contains(o.getClass()));
+        if (allNumbers) {
+            final long[] elements = list.stream().mapToLong(o -> ((Number) o).longValue()).toArray();
+            return new AbstractList<Object>() {
+                @Override
+                public Object get(int index) {
+                    return elements[index];
+                }
+                @Override
+                public int size() {
+                    return elements.length;
+                }
+            };
+        }
+
+        final boolean allStrings = list.stream().allMatch(o -> o != null && STRING_TYPES.contains(o.getClass()));
+        if (allStrings) {
+            final BytesRefBuilder builder = new BytesRefBuilder();
+            try (BytesStreamOutput bytesOut = new BytesStreamOutput()) {
+                final int[] endOffsets = new int[list.size()];
+                int i = 0;
+                for (Object o : list) {
+                    BytesRef b;
+                    if (o instanceof BytesRef) {
+                        b = (BytesRef) o;
+                    } else if (o instanceof CharBuffer) {
+                        b = new BytesRef((CharBuffer) o);
+                    } else {
+                        builder.copyChars(o.toString());
+                        b = builder.get();
+                    }
+                    bytesOut.writeBytes(b.bytes, b.offset, b.length);
+                    if (i == 0) {
+                        endOffsets[0] = b.length;
+                    } else {
+                        endOffsets[i] = Math.addExact(endOffsets[i-1], b.length);
+                    }
+                    ++i;
+                }
+                final BytesReference bytes = bytesOut.bytes();
+                return new AbstractList<Object>() {
+                    @Override
+                    public Object get(int i) {
+                        final int startOffset = i == 0 ? 0 : endOffsets[i-1];
+                        final int endOffset = endOffsets[i];
+                        return bytes.slice(startOffset, endOffset - startOffset).toBytesRef();
```

Comment:
Similarly here. Let's keep things in a `BytesReference`.

Reply:
@DaveCTurner  This is just some code existed originally, and need to be remove in a follow up. so maybe we do not need to optimize this? :)

Reply:
You can find more context [here](https://github.com/elastic/elasticsearch/pull/67223#discussion_r554591349)

Reply:
Ok I see. These ones are small so will mostly avoid copying.